pwndbg/aglib/vmmap_custom.py

```diff
@@ -0,0 +1,154 @@
+# by analyzing the stack or register context.
+explored_pages: List[pwndbg.lib.memory.Page] = []
+
+# List of custom pages that can be managed manually by vmmap_* commands family
+custom_pages: List[pwndbg.lib.memory.Page] = []
+
+auto_explore = pwndbg.config.add_param(
+    "auto-explore-pages",
+    "warn",
+    "whether to try to infer page permissions when memory maps missing (can cause errors)",
+    param_class=pwndbg.lib.config.PARAM_ENUM,
+    enum_sequence=["yes", "warn", "no"],
+)
+
+
+_warn_cache: Set[int] = set()
+
+
+@pwndbg.dbg.event_handler(EventType.NEW_MODULE)
+def clear_warn_cache():
+    _warn_cache.clear()
+
+
+def get_custom_pages() -> Tuple[pwndbg.lib.memory.Page, ...]:
+    """
+    Returns a tuple of `Page` objects representing the memory mappings of the
+    target, sorted by virtual address ascending.
+    """
+    pages: List[pwndbg.lib.memory.Page] = [*explored_pages, *custom_pages]
+    pages.sort()
+    return tuple(pages)
+
+
+def add_custom_page(page: pwndbg.lib.memory.Page) -> None:
+    bisect.insort(custom_pages, page)
+
+    # Reset all the cache
+    # We can not reset get() only, since the result may be used by others.
+    # TODO: avoid flush all caches
+    pwndbg.lib.cache.clear_caches()
+
+
+def clear_custom_page() -> None:
+    while custom_pages:
+        custom_pages.pop()
+
+    # Reset all the cache
+    # We can not reset get() only, since the result may be used by others.
+    # TODO: avoid flush all caches
+    pwndbg.lib.cache.clear_caches()
+
+
+def explore(address_maybe: int) -> pwndbg.lib.memory.Page | None:
+    """
+    Given a potential address, check to see what permissions it has.
+
+    Returns:
+        Page object
+
+    Note:
+        Adds the Page object to a persistent list of pages which are
+        only reset when the process dies. This means pages which are
+        added this way will not be removed when unmapped.
+
+        Also assumes the entire contiguous section has the same permission.
+    """
+    if not pwndbg.dbg.selected_inferior().is_linux():
+        return None
+
+    if auto_explore.value == "warn":
+        page_start = pwndbg.lib.memory.page_align(address_maybe)
+        if page_start not in _warn_cache:
+            _warn_cache.add(page_start)
+            is_readable_addr = pwndbg.aglib.memory.peek(page_start)
+            if is_readable_addr:
+                print(
+                    M.warn(
+                        f"Warning: Avoided exploring possible address {address_maybe:#x}.\n"
+                        f"You can explicitly explore it with `vmmap_explore {page_start:#x}`"
+                    )
+                )
+        return None
+    elif auto_explore.value == "no":
+        return None
+
+    address_maybe = pwndbg.lib.memory.page_align(address_maybe)
+    flags = get_memory_flags(address_maybe)
+    if flags is None:
+        return None
+
+    page = find_boundaries(address_maybe)
+    page.objfile = "<explored>"
```

Comment:
I wonder if this should say sth like `<explored:0x123>` where 0x123 is like idk, first X nibbles of an address

Reply:
done